docs/spec/auth/vesting.md

```diff
@@ -1,159 +1,463 @@
-    ReceivedCoins        sdk.Coins // Coins received from other accounts
-    SentCoins            sdk.Coins // Coins sent to other accounts
+    BaseVestingAccount
 
-    // StartTime and EndTime used to calculate how much of OriginalCoins is unlocked at any given point
-    StartTime time.Time
-    EndTime   time.Time
+    DelegatedVesting Coins // coins that vesting and delegated
+    StartTime        Time // when the coins start to vest
 }
 
-// Uses time in context to calculate total unlocked coins
-SendableCoins(vacc ContinuousVestingAccount, ctx sdk.Context) sdk.Coins:
-    
-    // Coins unlocked by vesting schedule
-    unlockedCoins := ReceivedCoins - SentCoins + OriginalVestingCoins * (Now - StartTime) / (EndTime - StartTime)
+// DelayedVestingAccount implements the VestingAccount interface. It vests all
+// coins after a specific time, but non prior. In other words, it keeps them
+// locked until a specified time.
+type DelayedVestingAccount struct {
+    BaseAccount
+    BaseVestingAccount
+}
+```
 
-    // Must still check for currentCoins constraint since some unlocked coins may have been delegated.
-    currentCoins := vacc.BaseAccount.GetCoins()
+## Vesting Account Specification
 
-    // min will return sdk.Coins with each denom having the minimum amount from unlockedCoins and currentCoins
-    return min(unlockedCoins, currentCoins)
+Given a vesting account, we define the following in the proceeding operations:
+
+- `OV`: The original vesting coin amount. It is a constant value.
+- `V`: The number of `OV` coins that are still _vesting_. It is derived by `OV`, `StartTime` and `EndTime`. This value is computed on demand and not on a per-block basis.
+- `V'`: The number of `OV` coins that are _vested_ (unlocked). This value is computed on demand and not a per-block basis.
+- `DV`: The number of delegated _vesting_ coins. It is a variable value. It is stored and modified directly in the vesting account.
+- `DF`: The number of delegated _vested_ (unlocked) coins. It is a variable value. It is stored and modified directly in the vesting account.
+- `BC`: The number of `OV` coins less any coins that are transferred, which can be negative, or delegated (`DV + DF`). It is considered to be balance of the embedded base account. It is stored and modified directly in the vesting account.
+
+### Determining Vesting & Vested Amounts
+
+It is important to note that these values are computed on demand and not on a
+mandatory per-block basis.
+
+#### Continuously Vesting Accounts
+
+To determine the amount of coins that are vested for a given block `B`, the
+following is performed:
+
+1. Compute `X := B.Time - StartTime`
+2. Compute `Y := EndTime - StartTime`
+3. Compute `V' := OV * (X / Y)`
+4. Compute `V := OV - V'`
+
+Thus, the total amount of _vested_ coins is `V'` and the remaining amount, `V`,
+is _vesting_.
+
+```go
+func (cva ContinuousVestingAccount) GetVestedCoins(b Block) Coins {
```

Comment:
I think we need to be a bit careful here and check for negatives - we may want to set a start time slightly in the future in case the time when a blockchain will launch is not exactly known - if the current time is less than the vesting start time, we should return zero.

Reply:
>  if the current time is less than the vesting start time, we should return zero.

This is a good overall sanity check regardless -- thanks 👍